src/main/kotlin/org/jetbrains/research/libsl/nodes/functions.kt

```diff
@@ -33,27 +41,142 @@ data class Function(
             append(returnType.resolve()?.fullName ?: UNRESOLVED_TYPE_SYMBOL)
         }
 
-        if (contracts.isNotEmpty()) {
-            appendLine()
-            append(withIndent(formatListEmptyLineAtEndIfNeeded(contracts)))
+        if (!hasBody && contracts.isEmpty()) {
+            appendLine(";")
+        } else if (hasBody) {
+            appendLine(" {")
+            if (contracts.isNotEmpty()) {
+                append(withIndent(formatListEmptyLineAtEndIfNeeded(contracts)))
+            }
+            append(withIndent(formatListEmptyLineAtEndIfNeeded(statements)))
+            appendLine("}")
+        }
+    }
+}
+
+data class Constructor(
+    var name: String?,
+    var args: MutableList<FunctionArgument> = mutableListOf(),
+    val annotationReferences: MutableList<AnnotationReference>? = mutableListOf(),
+    var contracts: MutableList<Contract> = mutableListOf(),
+    var statements: MutableList<Statement> = mutableListOf(),
+    var localVariables: MutableList<Variable> = mutableListOf(),
+    val hasBody: Boolean = statements.isNotEmpty(),
+    val context: FunctionContext
+) : Node() {
+    override fun dumpToString(): String = buildString {
+        annotationReferences?.joinToString() { annotationReference ->
+            append(annotationReference.resolveOrError().invocationDumpToString())
+        }
+
+        name = if(name.isNullOrEmpty()) {
+            ""
+        } else {
+            " " + BackticksPolitics.forIdentifier(name!!)
+        }
+
+        append("constructor$name")
+
+        append(
+            args.joinToString(separator = ", ", prefix = "(", postfix = ")") { arg -> arg.dumpToString() }
+        )
+
+        if (!hasBody && contracts.isEmpty()) {
+            appendLine(";")
+        } else if (hasBody) {
+            appendLine(" {")
+            if (contracts.isNotEmpty()) {
+                append(withIndent(formatListEmptyLineAtEndIfNeeded(contracts)))
+            }
+            append(withIndent(formatListEmptyLineAtEndIfNeeded(statements)))
+            appendLine("}")
+        }
+    }
+
+}
+
+data class Destructor(
```

Comment:
Same issues

Also, there is some code duplication. You must use OOP here

Reply:
I think that constructors and destructors (and may be procedures) could be described with one node class -- function. 

Idk about contracts on constructors/destructors but we can shift the responsibility for this to checkers that could be implemented later

Reply:
Done